In [1]:
import pandas as pd
import numpy as np
import requests
import json
import urllib
import datetime

pd.set_option("display.max_columns", 300)
pd.set_option('display.max_rows', 300)
pd.options.mode.chained_assignment = None

Load in `df_bq`, which contains results from [this query](https://bigquery.cloud.google.com/results/infusionsoft-data-production:US.bquijob_47811bc0_16253f5e66f) from our payment records, and `df_wp`, which contains an export of WePay CAD transaction data.

Then, limit `df_bq` and `df_wp` to payments in January and February, filter by app name `qv200`, and limit `df_bq` to `PaymentSubType = 'INF_PAY'`.

In [2]:
df_bq = pd.read_csv('results-20180323-104407.csv')
df_bq['day_created'] = pd.to_datetime(df_bq.DateCreated)
df_bq['day_created'] = pd.DatetimeIndex(df_bq['day_created'] ).normalize()

df_bq = df_bq[(df_bq.appName == 'qv200') & (df_bq.day_created >= '2018-01-01') & (df_bq.day_created < '2018-03-01') & (df_bq.PaymentSubType == 'INF_PAY')].reset_index()

df_bq

index      Id  ContactId  PaymentId  TotalAmountPaid  \
0       5  114251      14061       3505       226.000000   
1       7  112751       7711       3457       110.680000   
2       8  112749       7711       3455        33.840000   
3      14  113517      12315       3467        33.840000   
4      15  113519      20317       3469        67.690002   
5      17  120431      22689       3655        75.709999   
6      20  115935      12473       3565       110.680000   
7      21  115825      21519       3561        33.840000   
8      22  111929       3985       3433       110.680000   
9      23  120433      22055       3657        75.709999   
10     26  114071      20427       3481        33.840000   
11     27  114077      20431       3487       110.680000   
12     28  114075      11991       3485        67.690002   
13     32  118521       7711       3623       339.000000   
14     36  112221      20129       3437        33.840000   
15     39  120191      20673       3651       339.000000   
16     40  120189      22641       3649       565.000000   
17     41  117993      15161       3615       565.000000   
18     42  118075      22407       3617       565.000000   
19     45  120817      22883       3661       197.750000   
20     47  115045      20789       3537       226.000000   
21     52  114577       4663       3511        33.840000   
22     53  114663      20667       3517        33.840000   
23     54  114665      14301       3519        67.690002   
24     55  114619      20657       3513        33.840000   
25     56  120663       9961       3659       339.000000   
26     58  117635      20691       3609       339.000000   
27     60  112565      14225       3451       110.680000   
28     62  116305       7711       3575        33.840000   
29     63  116387      20789       3581        33.840000   
30     64  116385      20789       3579       110.680000   
31     65  116301       7711       3573        33.840000   
32     68  116589      11795       3585       339.000000   
33     72  116031      21721       3569        33.840000   
34     73  118237      22563       3619       197.750000   
35     75  119715      20159       3637       339.000000   
36     76  119757      14225       3641       339.000000   
37     77  119675      22589       3635       339.000000   
38     78  119673       6018       3633        75.709999   
39     80  117399      22491       3595       565.000000   
40     81  117401      22493       3597       197.750000   
41     82  112969       6319       3459       226.000000   
42     83  120027      22617       3645       339.000000   
43     86  114831      20703       3525        33.900002   
44     87  114829      20691       3523       226.000000   
45     88  114787      12327       3521       110.680000   
46     89  120971      11499       3663       565.000000   
47     90  115495      21265       3547        33.840000   
48     91  115497      21271       3549       565.000000   
49     95  113115      20237       3461        33.840000   
50     96  113189      20269       3463       420.359985   
51     97  119911       6319       3643       339.000000   
52     99  119249      14225       3631       151.419998   
53    100  119205      21323       3627       565.000000   
54    101  112111      18569       3435       197.750000   
55    102  116491      22405       3583       110.680000   
56    103  113629       6675       3471       110.680000   
57    104  113631      20269       3473       565.000000   
58    105  113633      20343       3475       339.000000   
59    106  117475      22511       3599       565.000000   
60    116  115703      15585       3557       226.000000   
61    119  121329      20405       3673       197.750000   
62    120  121327      23343       3671       339.000000   
63    121  121285      20405       3669       197.750000   

                       PaymentDate TransactionType CollectionMethod  \
0   2018-01-16 16:51:27.000000 U

In [3]:
df_wp = pd.read_csv('wepay.csv')
df_wp['create_time'] = pd.to_datetime(df_wp.create_time, unit='s')
df_wp['day_created'] = pd.to_datetime(df_wp['create_time'])
df_wp['day_created'] = pd.DatetimeIndex(df_wp['day_created'] ).normalize()

df_wp = df_wp[df_wp.account_name.str.contains('qv200') & (df_wp.day_created >= '2018-01-01') & (df_wp.day_created < '2018-03-01')].reset_index()

df_wp.drop(['account_name','checkout_uri','payer_email','payer_name'], axis=1, inplace=True)

df_wp

index  checkout_id  account_id   type                 short_description  \
0    1112   1854219689   479437278  GOODS  Auto Charge For Invoice Ids:3377   
1    1131   1442245989   479437278  GOODS  Auto Charge For Invoice Ids:3375   
2    1244    693674592   479437278  GOODS  Auto Charge For Invoice Ids:3367   
3    1282   1624981774   479437278  GOODS  Auto Charge For Invoice Ids:3365   
4    1294    453644247   479437278  GOODS  Auto Charge For Invoice Ids:3363   
5    1328    353088085   479437278  GOODS  Auto Charge For Invoice Ids:3361   
6    1343   1207757106   479437278  GOODS  Auto Charge For Invoice Ids:3359   
7    1392    456973385   479437278  GOODS  Auto Charge For Invoice Ids:3357   
8    1394   2015559781   479437278  GOODS  Auto Charge For Invoice Ids:3355   
9    1435   1624983822   479437278  GOODS  Auto Charge For Invoice Ids:3351   
10   1468     88614843   479437278  GOODS  Auto Charge For Invoice Ids:3349   
11   1502    642048812   479437278  GOODS  Auto Charge For Invoice Ids:3347   
12   1507   1887684556   479437278  GOODS  Auto Charge For Invoice Ids:3345   
13   1508   1500331886   479437278  GOODS  Auto Charge For Invoice Ids:3343   
14   1638   1638277581   479437278  GOODS  Auto Charge For Invoice Ids:3341   
15   1650   1640410665   479437278  GOODS  Auto Charge For Invoice Ids:3337   
16   1786    894037644   479437278  GOODS  Auto Charge For Invoice Ids:3333   
17   1824   1339145853   479437278  GOODS  Auto Charge For Invoice Ids:3329   
18   1860   1502387185   479437278  GOODS  Auto Charge For Invoice Ids:3327   
19   1886    505882859   479437278  GOODS  Auto Charge For Invoice Ids:3325   
20   1970   1034923738   479437278  GOODS  Auto Charge For Invoice Ids:3315   
21   2021   1900516329   479437278  GOODS  Auto Charge For Invoice Ids:3307   
22   2035   1998131520   479437278  GOODS  Auto Charge For Invoice Ids:3305   
23   2041   1502742921   479437278  GOODS  Auto Charge For Invoice Ids:3303   
24   2318   1464710583   479437278  GOODS  Auto Charge For Invoice Ids:3291   
25   2343   1280937345   479437278  GOODS  Auto Charge For Invoice Ids:3289   
26   2359   1300487241   479437278  GOODS  Auto Charge For Invoice Ids:3287   
27   2368   2125847028   479437278  GOODS  Auto Charge For Invoice Ids:3285   
28   2378   1969011946   479437278  GOODS  Auto Charge For Invoice Ids:3283   
29   2433    246648361   479437278  GOODS  Auto Charge For Invoice Ids:3279   
30   2457   1030163417   479437278  GOODS  Auto Charge For Invoice Ids:3275   
31   2486   1575224609   479437278  GOODS  Auto Charge For Invoice Ids:3269   
32   2504   1216630397   479437278  GOODS  Auto Charge For Invoice Ids:3265   
33   2543    114993459   479437278  GOODS  Auto Charge For Invoice Ids:3259   
34   2551   2097360337   479437278  GOODS  Auto Charge For Invoice Ids:3257   
35   2637   1113320832   479437278  GOODS  Auto Charge For Invoice Ids:3251   
36   2669     77630722   479437278  GOODS  Auto Charge For Invoice Ids:3245   
37   2676    593688922   479437278  GOODS  Auto Charge For Invoice Ids:3243   
38   2687   1232953577   479437278  GOODS  Auto Charge For Invoice Ids:3241   
39   2711    195507927   479437278  GOODS  Auto Charge For Invoice Ids:3239   
40   2717   1957594970   479437278  GOODS  Auto Charge For Invoice Ids:3237   
41   2734    477369397   479437278  GOODS  Auto Charge For Invoice Ids:3233   
42   2739   1902291226   479437278  GOODS  Auto Charge For Invoice Ids:3231   
43   2837   1273604418   479437278  GOODS  Auto Charge For Invoice Ids:3225   
44   2948    776827504   479437278  GOODS  Auto Charge For Invoice Ids:3215   
45   2951    416612919   479437278  GOODS  Auto Charge For Invoice Ids:3213   
46   2956    515513228   479437278  GOODS  Auto Charge For Invoice Ids:3209   
47   3064   1976908902   479437278  GOODS  Auto Charge For Invoice Ids:3203   
48   3065    878383702   479437278  GOODS  Auto Charge For Invoice Ids:3201   
49   3076   1640892029   479437278  GOODS  Auto Charge

The lengths of these two identically-filtered datasets are given:

In [4]:
print(f'Our BQ data is {len(df_bq)} rows')
print(f'The WePay data is {len(df_wp)} rows')

Our BQ data is 64 rows
The WePay data is 61 rows


Group these two by day into `wp_groupdays` and `bq_groupdays`, and then find for which days the transaction counts are different.

In [5]:
wp_groupdays = df_wp.groupby('day_created').agg('count')['amount'].reset_index()
bq_groupdays = df_bq.groupby('day_created').agg('count')['TotalAmountPaid'].reset_index()

days_unmatched = []
for x in range(0,len(bq_groupdays)):
    if bq_groupdays.iloc[x,1] != wp_groupdays.iloc[x,1]:
        print(bq_groupdays.iloc[x,0])
        print(f'bq says we have {bq_groupdays.iloc[x,1]} transactions on this date, while wp says we have {wp_groupdays.iloc[x,1]}')
        days_unmatched.append(datetime.datetime.strftime(bq_groupdays.iloc[x,0], format='%Y-%m-%d'))

2018-01-28 00:00:00
bq says we have 2 transactions on this date, while wp says we have 1
2018-02-18 00:00:00
bq says we have 4 transactions on this date, while wp says we have 3
2018-02-28 00:00:00
bq says we have 3 transactions on this date, while wp says we have 2


Loop through these `unmatched_days` and print off what we have.

In [6]:
df_bq[df_bq.day_created.isin(days_unmatched)].sort_values('PaymentDate')

index      Id  ContactId  PaymentId  TotalAmountPaid  \
31     65  116301       7711       3573        33.840000   
28     62  116305       7711       3575        33.840000   
38     78  119673       6018       3633        75.709999   
37     77  119675      22589       3635       339.000000   
35     75  119715      20159       3637       339.000000   
36     76  119757      14225       3641       339.000000   
63    121  121285      20405       3669       197.750000   
62    120  121327      23343       3671       339.000000   
61    119  121329      20405       3673       197.750000   

                       PaymentDate TransactionType CollectionMethod  \
31  2018-01-28 15:34:34.000000 UTC         CAPTURE  PAYMENT_GATEWAY   
28  2018-01-28 15:35:42.000000 UTC          REFUND  PAYMENT_GATEWAY   
38  2018-02-18 14:12:23.000000 UTC         CAPTURE  PAYMENT_GATEWAY   
37  2018-02-18 14:50:58.000000 UTC         CAPTURE  PAYMENT_GATEWAY   
35  2018-02-18 16:20:43.000000 UTC          REFUND  PAYMENT_GATEWAY   
36  2018-02-18 23:19:08.000000 UTC         CAPTURE  PAYMENT_GATEWAY   
63  2018-02-28 14:39:39.000000 UTC         CAPTURE  PAYMENT_GATEWAY   
62  2018-02-28 20:10:26.000000 UTC         CAPTURE  PAYMENT_GATEWAY   
61  2018-02-28 20:14:25.000000 UTC          REFUND  PAYMENT_GATEWAY   

   PaymentSubType  PaymentGatewayId  ErrorMessage  TestPayment  \
31        INF_PAY                 1           NaN        False   
28        INF_PAY                 1           NaN        False   
38        INF_PAY                 1           NaN        False   
37        INF_PAY                 1           NaN        False   
35        INF_PAY                 1           NaN        False   
36        INF_PAY                 1           NaN        False   
63        INF_PAY                 1           NaN        False   
62        INF_PAY                 1           NaN        False   
61        INF_PAY                 1           NaN        False   

    GeneralValidationPassed  StrategyValidationPassed  ExecuteRan  \
31                     True                      True        True   
28                     True                      True        True   
38                     True                      True        True   
37                     True                      True        True   
35                     True                      True        True   
36                     True                      True        True   
63                     True                      True        True   
62                     True                      True        True   
61                     True                      True        True   

    ResultsBuilt  ErrorsMapped     Source  Deleted  \
31          True         False  orderForm    False   
28          True         False        NaN    False   
38          True         False  orderForm    False   
37          True         False  orderForm    False   
35          True         False        NaN    False   
36          True         False  orderForm    False   
63          True         False  orderForm    False   
62          True         False  orderForm    False   
61          True         False        NaN    False   

                       DateCreated                     LastUpdated appName  \
31  2018-01-28 15:34:37.000000 UTC  2018-01-28 15:34:37.000000 UTC   qv200   
28  2018-01-28 15:35:42.000000 UTC  2018-01-28 15:35:42.000000 UTC   qv200   
38  2018-02-18 14:12:28.000000 UTC  2018-02-18 14:12:28.000000 UTC   qv200   
37  2018-02-18 14:51:03.000000 UTC  2018-02-18 14:51:03.000000 UTC   qv200   
35  2018-02-18 16:20:45.000000 UTC  2018-02-18 16:20:45.000000 UTC   qv200   
36  2018-02-18 23:19:11.000000 UTC  2018-02-18 23:19:11.000000 UTC   qv200   
63  2018-02-28 14:39:44.000000 UTC  2018-02-28 14:39:44.000000 UTC   qv200   
62  2018-02-28 20:10:32.000000 UTC  2018-02-28 20:10:32.000000 UTC   qv200   
61  2018-02-28 20:14:27.000000 UTC  2018-02-28 20:14:27.000000 UTC   qv200   

        PayType Cu

In [7]:
df_wp[df_wp.day_created.isin(days_unmatched)].sort_values('create_time')

index  checkout_id  account_id   type                 short_description  \
28   2378   1969011946   479437278  GOODS  Auto Charge For Invoice Ids:3283   
13   1508   1500331886   479437278  GOODS  Auto Charge For Invoice Ids:3343   
12   1507   1887684556   479437278  GOODS  Auto Charge For Invoice Ids:3345   
11   1502    642048812   479437278  GOODS  Auto Charge For Invoice Ids:3347   
1    1131   1442245989   479437278  GOODS  Auto Charge For Invoice Ids:3375   
0    1112   1854219689   479437278  GOODS  Auto Charge For Invoice Ids:3377   

   currency  amount fee_payer     state  redirect_uri  auto_capture  app_fee  \
28      CAD   33.84     payee  refunded           NaN             1        0   
13      CAD   75.71     payee  captured           NaN             1        0   
12      CAD  339.00     payee  captured           NaN             1        0   
11      CAD  339.00     payee  captured           NaN             1        0   
1       CAD  197.75     payee  refunded           NaN             1        0   
0       CAD  339.00     payee  captured           NaN             1        0   

           create_time  amount_refunded  amount_charged_back   gross    fee  \
28 2018-01-28 15:34:34            33.84                  0.0   33.84   1.28   
13 2018-02-18 14:12:25             0.00                  0.0   75.71   2.49   
12 2018-02-18 14:51:00             0.00                  0.0  339.00  10.13   
11 2018-02-18 23:19:08             0.00                  0.0  339.00  10.13   
1  2018-02-28 14:39:41           197.75                  0.0  197.75   6.03   
0  2018-02-28 20:10:29             0.00                  0.0  339.00  10.13   

    long_description  reference_id  \
28               NaN           NaN   
13               NaN           NaN   
12               NaN           NaN   
11               NaN           NaN   
1                NaN           NaN   
0                NaN           NaN   

                                         callback_uri  shipping_address  \
28  https://qv200.infusionsoft.com/app/ipn/wepay/t...               NaN   
13  https://qv200.infusionsoft.com/app/ipn/wepay/t...               NaN   
12  https://qv200.infusionsoft.com/app/ipn/wepay/t...               NaN   
11  https://qv200.infusionsoft.com/app/ipn/wepay/t...               NaN   
1   https://qv200.infusionsoft.com/app/ipn/wepay/t...               NaN   
0   https://qv200.infusionsoft.com/app/ipn/wepay/t...               NaN   

    shipping_fee  preapproval_id  tax day_created  
28             0             NaN    0  2018-01-28  
13             0             NaN    0  2018-02-18  
12             0             NaN    0  2018-02-18  
11             0             NaN    0  2018-02-18  
1              0             NaN    0  2018-02-28  
0              0             NaN    0  2018-02-28

We can now see the three entries in BQ that are not in WePay - the CAPTURE for 33.84 on 2018-01-28, the REFUND for 339.00 on 2018-02-18, and the CAPTURE for 197.75 on 2018-02-28. 